tools/make.py

```diff
@@ -249,6 +253,13 @@
                            "Currently set search path: %s"
                            %(toolchain, search_path))
 
+    manifests = set()
+    for src_dir in options.source_dir:
+        for root, dirnames, filenames in os.walk(src_dir):
+            for filename in fnmatch.filter(filenames, '*_psa.json'):
+                manifests.add(join(root, filename))
+    generate_partitions_sources(list(manifests))
```

Comment:
This seems like a bad place to trigger this. Especially because this will lead to duplicated code in test.py as this applies to building _any_ application.

Reply:
We had to split the code generation into 2 phases.
The first generates partition specific code, and validates all the manifests present in the tree (including partition dependencies).
The second phase generates only the common data which is app/test specific.

the reasons for the change are:
1. the way tests are built
2. allows mbed-os to hold partitions within it
3. make the validation phase stronger, we can now detect issues with the manifests before any compilation has occurred.

in the future i would like to introduce new commands to mbed-cli to list all partitions/secure functions.

Reply:
@orenc17 Thanks for letting me know the requirements. At the very minimum, there should be a helper function that does this thing for both test and make.

If I may be a tad more picky, I don't like that export does this _after_ scanning, and compile/test does this _before_ scanning. As a step up, can we make it work the same way in both cases. Given your requirements, fail fast + strong validation, I see no reason to require it to be implemented inside what should soon become a less complex CLI argument parser.

Here is what I would like to see:
 * Helper function in `tools/spm` to glue it into the build system (uses resources object or takes the source dict/list directly)
 * `build_library` and `bulid_application` both call this helper function themselves, or it's part of a common call, like bootloader mode and `toolchain.set_config_data`.
 * `build_library` or the `toolchain.copy_files` method know not to copy the SPM specific files, so that generation happens at most once. This should allow you to use the generated files in the tests, without recreating them every time.

If you would like, I can prototype this on a branch of my own, and we can discuss the merits of my suggestion vs your implementation.

Reply:
Created a helper function for test/make.
regarding export, since we cant export tests there is no need to scan for manifests in early stage.
export finds all the manifests it needs from the resources object and by that prevents another scan of the tree

Reply:
If export can find all of the manifests from the resource object the so can compile and test.

Reply:
> prevents another scan of the tree

We should only ever need one scan, though that scan may be incremental.